### DISCUSSION TITLE
How to Write Integration Tests for .NET Class Library Without Moq or Dummy Test Classes?

### DISCUSSION THREAD

**Discussion by GevorDanielyan** [xunit/xunit#2898] [Question] [answered]
I'm currently developing a .NET class library that provides maintenance mode functionality for ASP.NET Core applications. I want to ensure the reliability and correctness of my library by writing integration tests, but I'm struggling to find resources on how to do this effectively without relying on mocking frameworks like Moq or creating dummy test classes for my services.

My library includes services that interact with external dependencies, middleware for handling maintenance mode requests, and controller actions for enabling/disabling maintenance mode. I want to validate the interactions between these components and ensure that the library behaves as expected in a real-world scenario.

Could someone provide guidance or best practices on how to approach writing integration tests for a .NET class library without using mocking frameworks or dummy test classes? I'm particularly interested in strategies for setting up a test environment, configuring dependencies, and testing interactions between components in an integrated manner.

Any insights, examples, or recommended resources would be greatly appreciated. Thank you!

**Comment by bradwilson** [comment, marked as answer]
> I'm struggling to find resources on how to do this effectively without relying on mocking frameworks like Moq or creating dummy test classes for my services.

Mocking frameworks and dummy classes tend to be more for unit testing than integration testing, but either way they serve an important purpose: to replace something that does work with something that doesn't do work, so that you can inspect what was asked to be done. It's basically an impossible to task to ask for something to be programmatically testable and yet not change/replace any of the production code to do so.

Here are a couple options you can use:

- Run the application. Hit it with a "fake browser" and look at the results to see if you get what you expected.
- Add observability to the application. Run it. Hit it with a real browser. Inspect the resulting telemetry to ensure that you can verify that what you expected to happen, happened.

These are in addition to the more common option of using [`TestServer`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.testhost.testserver), which I'm assuming you'd reject as being a fake (and you're trying to avoid fakes).

Also, it's important to note that what you have here is not an application, it's a class library. There's no such thing as "end to end testing" for a class library, because it's made up of components and not an application. Class libraries are usually primarily tested via unit tests, and unit tests primarily become possible because of fakes that are placed at the boundaries of what you want to test. I'm unsure why you're unhappy using fakes, but that's the norm. You may be fighting against something that's not reasonable to fight against for the kind of testing you need to do.

**Comment by mwpowellhtx** [comment]
Good looking dates _Mar 2024_, so I am not the only one in recent memory to want to do something like this. I am testing an API client that I am writing, but the thing is, there are some critical bits coming from a login sequence of the API server. I need to be able to bare minimum initiate that sequence, and listen for outgoing HTTPS requests. After I have that bit of detail, then. Ideally also, should be runtime friendly, not necessarily test, compatible, `Sdk`-wise. Maybe the better thing to do is to actually host a web page control, perhaps in a WPF application, and see about intercepting requests from that perspective. I was just hoping there might be something I could do from a rather simple `HttpClient` only based approach, but I am finding that it lacks the runtime browser awareness. So no page forwards occur, nothing like that, I just get the static one time response, which is a start. But there are other RESTful calls going on which I do require. Thoughts?
